src/main/java/com/hivemq/mqtt/message/pool/FreeIdRanges.java

```diff
@@ -0,0 +1,198 @@
+ * and to reduce the memory footprint of keeping track of allocated IDs.
+ * <p>
+ * This is achieved by keeping a list of {@link Range} objects each of which represents a contiguous interval of
+ * integer ids that are NOT currently assigned to any object using this instance of {@link FreeIdRanges}.
+ * The lower end of the {@link Range} instance is included whereas the upper end is excluded from the interval.
+ * That is, the {@link Range} instance with start 42 and end 42 contains only one ID - 42.
+ * The {@link Range} instance with start 10 and end 12 contains only two IDs: 10 and 11.
+ * <p>
+ * Initially, there is only one contiguous {@link Range} of IDs. The IDs are assigned starting from the lowest one in
+ * that interval.
+ * Upon assignment, the interval's lower end is incremented (interval size reduces from below).
+ * When the ID is returned, it either joins one of the existing {@link Range} intervals in the list (if it is adjacent)
+ * or it forms a new {@link Range} that is added to the list.
+ * <p>
+ * This class is NOT thread-safe.
+ * <a
+ * href="https://github.com/hivemq/hivemq-mqtt-client/blob/master/src/main/java/com/hivemq/client/internal/util/Ranges.java">The
+ * original implementation in the HiveMQ Java Client.</a>
+ */
+public class FreeIdRanges {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(FreeIdRanges.class);
+
+    final int minAllowedId;
+    final int maxAllowedId;
+
+    private @NotNull Range rootRange;
+
+    public FreeIdRanges(final int minId, final int maxId) {
+        minAllowedId = minId;
+        maxAllowedId = maxId;
+        rootRange = new Range(minId, maxId + 1);
+    }
+
+    /**
+     * Provides a new ID that is not currently allocated.
+     *
+     * @return a new ID if available in any of the ranges or {@link NoMessageIdAvailableException} if ran out of IDs.
+     */
+    public int takeNextId() throws NoMessageIdAvailableException {
+        if (rootRange.start == rootRange.end) {
+            throw new NoMessageIdAvailableException();
+        }
+
+        final int id = rootRange.start;
+        rootRange.start++;
+        if ((rootRange.start == rootRange.end) && (rootRange.next != null)) {
+            rootRange = rootRange.next;
+        }
+        return id;
+    }
+
+    /**
+     * Provides the requested ID if it is available or some other ID otherwise.
+     *
+     * @param id an ID that the caller attempts to take.
+     * @return the requested {@param id} if it is available in one of the {@link Range} intervals or otherwise some
+     *         other free ID.
+     */
+    public int takeIfAvailable(final int id) throws NoMessageIdAvailableException {
+        if (id < minAllowedId || id > maxAllowedId) {
+            log.warn("Attempting to take an ID {} that is outside the valid range [{}, {}], will try taking another ID.",
+                    id,
+                    minAllowedId,
+                    maxAllowedId);
+        }
```

Comment:
This is a programming error, so imo we should use `Preconditions.checkArgument` here.